### ISSUE TITLE
ciscoasa, syntax error

### ISSUE THREAD

**Issue by t3chn0m4g3** [vorband/ewsposter#12]
when parsing ciscoasa logs, ewsposter crashes with the following error

```
Traceback (most recent call last):
  File "/opt/ewsposter/ews.py", line 2423, in <module>
    eval(i+'()')
  File "<string>", line 1, in <module>
  File "/opt/ewsposter/ews.py", line 2218, in ciscoasa
    linecontent=ast.literal_eval(line)
  File "/usr/lib/python2.7/ast.py", line 49, in literal_eval
    node_or_string = parse(node_or_string, mode='eval')
  File "/usr/lib/python2.7/ast.py", line 37, in parse
    return compile(source, filename, mode, PyCF_ONLY_AST)
  File "<unknown>", line 1
    {'timestamp': '2018-07-09T20:36:31.796618', 'src_ip': '71.6.146.185', 'payload_printable': 'Request timed out: timeout('The read operation timed out',)'}
                                                                                                                              ^
SyntaxError: invalid syntax
```

Log in question

```
Starting server on port 5000/udp
Using selector: EpollSelector
Starting server on port 8443/tcp, use <Ctrl-C> to stop
{'timestamp': '2018-07-09T20:36:28.792805', 'src_ip': '71.6.146.185', 'payload_printable': '"GET / HTTP/1.1" 200 -'}
{'timestamp': '2018-07-09T20:36:31.796618', 'src_ip': '71.6.146.185', 'payload_printable': 'Request timed out: timeout('The read operation timed out',)'}
----------------------------------------
Exception happened during processing of request from ('71.6.146.185', 39674)
Traceback (most recent call last):
  File "/usr/lib/python3.6/socketserver.py", line 317, in _handle_request_noblock
    self.process_request(request, client_address)
  File "/usr/lib/python3.6/socketserver.py", line 348, in process_request
    self.finish_request(request, client_address)
  File "/usr/lib/python3.6/socketserver.py", line 361, in finish_request
    self.RequestHandlerClass(request, client_address, self)
  File "/usr/lib/python3.6/socketserver.py", line 696, in __init__
    self.handle()
  File "/usr/lib/python3.6/http/server.py", line 418, in handle
    self.handle_one_request()
  File "asa_server.py", line 193, in handle_one_request
    self.raw_requestline = self.rfile.readline(65537)
  File "/usr/lib/python3.6/socket.py", line 586, in readinto
    return self._sock.recv_into(b)
  File "/usr/lib/python3.6/ssl.py", line 1009, in recv_into
    return self.read(nbytes, buffer)
  File "/usr/lib/python3.6/ssl.py", line 871, in read
    return self._sslobj.read(len, buffer)
  File "/usr/lib/python3.6/ssl.py", line 631, in read
    v = self._sslobj.read(len, buffer)
ssl.SSLError: [SSL: UNEXPECTED_RECORD] unexpected record (_ssl.c:2217)
----------------------------------------
{'timestamp': '2018-07-09T20:36:40.999808', 'src_ip': '71.6.146.185', 'payload_printable': '"GET /robots.txt HTTP/1.1" 404 -'}
{'timestamp': '2018-07-09T20:36:41.657481', 'src_ip': '71.6.146.185', 'payload_printable': '"GET /sitemap.xml HTTP/1.1" 404 -'}
----------------------------------------
Exception happened during processing of request from ('71.6.146.185', 41001)
Traceback (most recent call last):
  File "/usr/lib/python3.6/socketserver.py", line 317, in _handle_request_noblock
    self.process_request(request, client_address)
  File "/usr/lib/python3.6/socketserver.py", line 348, in process_request
    self.finish_request(request, client_address)
  File "/usr/lib/python3.6/socketserver.py", line 361, in finish_request
    self.RequestHandlerClass(request, client_address, self)
  File "/usr/lib/python3.6/socketserver.py", line 696, in __init__
    self.handle()
  File "/usr/lib/python3.6/http/server.py", line 420, in handle
    self.handle_one_request()
  File "asa_server.py", line 193, in handle_one_request
    self.raw_requestline = self.rfile.readline(65537)
  File "/usr/lib/python3.6/socket.py", line 586, in readinto
    return self._sock.recv_into(b)
  File "/usr/lib/python3.6/ssl.py", line 1009, in recv_into
    return self.read(nbytes, buffer)
  File "/usr/lib/python3.6/ssl.py", line 871, in read
    return self._sslobj.read(len, buffer)
  File "/usr/lib/python3.6/ssl.py", line 631, in read
    v = self._sslobj.read(len, buffer)
ConnectionResetError: [Errno 104] Connection reset by peer
----------------------------------------
{'timestamp': '2018-07-09T20:36:42.620603', 'src_ip': '71.6.146.185', 'payload_printable': '"GET /.well-known/security.txt HTTP/1.1" 404 -'}
----------------------------------------
Exception happened during processing of request from ('71.6.146.185', 41201)
Traceback (most recent call last):
  File "/usr/lib/python3.6/socketserver.py", line 317, in _handle_request_noblock
    self.process_request(request, client_address)
  File "/usr/lib/python3.6/socketserver.py", line 348, in process_request
    self.finish_request(request, client_address)
  File "/usr/lib/python3.6/socketserver.py", line 361, in finish_request
    self.RequestHandlerClass(request, client_address, self)
  File "/usr/lib/python3.6/socketserver.py", line 696, in __init__
    self.handle()
  File "/usr/lib/python3.6/http/server.py", line 420, in handle
    self.handle_one_request()
  File "asa_server.py", line 193, in handle_one_request
    self.raw_requestline = self.rfile.readline(65537)
  File "/usr/lib/python3.6/socket.py", line 586, in readinto
    return self._sock.recv_into(b)
  File "/usr/lib/python3.6/ssl.py", line 1009, in recv_into
    return self.read(nbytes, buffer)
  File "/usr/lib/python3.6/ssl.py", line 871, in read
    return self._sslobj.read(len, buffer)
  File "/usr/lib/python3.6/ssl.py", line 631, in read
    v = self._sslobj.read(len, buffer)
ConnectionResetError: [Errno 104] Connection reset by peer
----------------------------------------
{'timestamp': '2018-07-09T20:36:44.201870', 'src_ip': '71.6.146.185', 'payload_printable': '"GET /favicon.ico HTTP/1.1" 404 -'}

```

**Comment by vorband** [comment]
Just to be sure, the log you provided does not produce any submission, as no `src_port` field is available, see https://github.com/vorband/ewsposter/issues/8. But of course the error must be caught.

The error is provoked by the ciscoasa honeypot which is generating invalid json. 

By [definition](http://www.json.org/), json must use double quotes, not single quotes. I already managed to work around this by using ast.literal_eval(). 

In this particular example, the content of `payload_printable` also includes single quotes, which breaks the processing.

` {
   'timestamp': '2018-07-09T20:36:31.796618', 
   'src_ip': '71.6.146.185', 
   'payload_printable': 'Request timed out: timeout('The read operation timed out',)'
}`

The value for `payload_printable` is now `'Request timed out: timeout('`. The following `The read operation timed out',)'` is invalid and causes the crash.

I managed to work around this, by ignoring the lines that are invalid, so ewsposter won't crash. However, the root of the problem needs to be fixed on ciscoasa-side.
